SourceCode/GPS/Forms/FormSaving.cs

```
@@ -0,0 +1,73 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Data;
+using System.Drawing;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Windows.Forms;
+
+namespace AgOpenGPS
+{
+    public partial class FormSaving : Form
+    {
+        public FormSaving()
+        {
+            InitializeComponent();
+        }
+
+        public void InitializeSteps(bool isJobStarted)
+        {
+            lstSteps.Items.Clear();
+
+            if (isJobStarted)
+            {
+                lstSteps.Items.Add(ShutdownSteps.SaveParams);
+                lstSteps.Items.Add(ShutdownSteps.SaveField);
+                lstSteps.Items.Add(ShutdownSteps.SaveSettings);
+                lstSteps.Items.Add(ShutdownSteps.Finalizing);
+            }
+            else
+            {
+                lstSteps.Items.Add(ShutdownSteps.SaveSettings);
+                lstSteps.Items.Add(ShutdownSteps.Finalizing);
+            }
+        }
+
+        public void UpdateStep(int index, string text)
+        {
+            if (index >= 0 && index < lstSteps.Items.Count)
```

I think this range check is not what you want in this case.
The `index` is always predetermined and if it is not in the valid range, it should crash the application because it is a programming error.

---

Because we change the index due to AgShare enabled or not

---

I know, but even with that it would be better to crash the program and tell the developer: "hey you messed up your logic" than just silently do nothing (which may result in hard to reproduce bug reports).

---

I will change in the next commit.. good one